bib/pkg/progress/progress.go

```diff
@@ -311,29 +312,64 @@ func (b *debugProgressBar) SetProgress(subLevel int, msg string, done int, total
 	return nil
 }
 
-// XXX: merge variant back into images/pkg/osbuild/osbuild-exec.go
-func RunOSBuild(pb ProgressBar, manifest []byte, store, outputDirectory string, exports, extraEnv []string) error {
+type OSBuildOptions struct {
+	StoreDir  string
+	OutputDir string
+	ExtraEnv  []string
+
+	// BuildLog writes the osbuild output to the given writer
+	BuildLog io.Writer
+}
+
+func writersForOSBuild(pb ProgressBar, opts *OSBuildOptions, internalBuildLog io.Writer) (osbuildStdout io.Writer, osbuildStderr io.Writer) {
 	// To keep maximum compatibility keep the old behavior to run osbuild
 	// directly and show all messages unless we have a "real" progress bar.
 	//
 	// This should ensure that e.g. "podman bootc" keeps working as it
-	// is currently expecting the raw osbuild output. Once we double
-	// checked with them we can remove the runOSBuildNoProgress() and
-	// just run with the new runOSBuildWithProgress() helper.
+	// is currently expecting the raw osbuild output.
 	switch pb.(type) {
-	case *terminalProgressBar, *debugProgressBar:
-		return runOSBuildWithProgress(pb, manifest, store, outputDirectory, exports, extraEnv)
+	case *verboseProgressBar:
+		// No external build log requested and we won't need an
+		// internal one because all output goes directly to
+		// stdout/stderr. This is for maximum compatibility with
+		// the existing bootc-image-builder in "verbose" mode
+		// where stdout, stderr come directly from osbuild.
+		if opts.BuildLog == nil {
+			return osStdout, osStderr
+		}
+		// With a build log we need a single output stream
+		osbuildStdout = osStdout
 	default:
-		return runOSBuildNoProgress(pb, manifest, store, outputDirectory, exports, extraEnv)
+		// hide the direct osbuild output by default
+		osbuildStdout = io.Discard
 	}
-}
 
-func runOSBuildNoProgress(pb ProgressBar, manifest []byte, store, outputDirectory string, exports, extraEnv []string) error {
-	_, err := osbuild.RunOSBuild(manifest, store, outputDirectory, exports, nil, extraEnv, false, os.Stderr)
-	return err
+	// There is a slight wrinkle here: when requesting a buildlog
+	// we can no longer write to separate stdout/stderr streams
+	// without being racy and give potential out-of-order output
+	// (which is very bad and confusing in a log). The reason is
+	// that if cmd.Std{out,err} are different "go" will start two
+	// go-routine to monitor/copy those are racy when both stdout,stderr
+	// output happens close together (TestRunOSBuildWithBuildlog demos
+	// that). We cannot have our cake and eat it so here we need to
+	// combine osbuilds stderr into our stdout.
```

Comment:
Just a question but could we turn off buffering and how both goros write to a queue that we then write into the log or are there bears on that road too?

Reply:
I tried a few strategies, including writing a syncedMultiWriter that uses a mutex to ensure ordering but I couldn't get it race-free and the test would show out of sync output every 1/4 or of the time on my system. It is probably worth another attempt but I'm not super hopeful. I can think more about it.

Reply:
We can skip this bit for now since there's something working, right? :)

Reply:
Yes, it does work reliably and arguably redirecting stderr from osbuild directly to the users stderr is a bit of a bug anyway so I'm okay with this. Plus for the "verbose" case the old behavior is kept (no old scripts or integrations that depend on stderr will break)